### PR TITLE
Update Throws to use ClientLogger.logAndThrow

### PR DESCRIPTION
**Pull request by alzimmermsft**
Updates all instance based throws to use ClientLogger.logAndThrow instead as per our spec. Additionally, there was some changes to the exception types being thrown to better align with the spec as well.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 1]
```diff
@@ -152,7 +155,8 @@ public URL getBlobUrl() {
             }
             return urlBuilder.toURL();
         } catch (MalformedURLException e) {
-            throw new RuntimeException(String.format("Invalid URL on %s: %s" + getClass().getSimpleName(), azureBlobStorage.getUrl()), e);
+            logger.logAndThrow(new IllegalStateException(String.format("Invalid URL on %s: %s" + getClass().getSimpleName(), azureBlobStorage.getUrl()), e));
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: It throws different exception in different client.
It does not make sense to throw IllegalArgument as it is not an argument here. 
IllegalState is using for the following case:
> Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation.

Can we return the RuntimeException? What's the concern of throwing RuntimeException?

**Comment by joshfree** [comment]
We have a lot of Storage PRs right now.  @mssfang and @sima-zhu could you review this PR with @alzimmermsft?  After our local team has done the review / iteration / updates, @alzimmermsft can then circle back with @rickle-msft or @jaschrep-msft for a final sanity check.

**Comment by sima-zhu** [comment]
Hello Josh,

Rick just replied this PR wouldn’t happen. Do we still need to active review and check in this?

**Comment by joshfree** [comment]
@sima-zhu yes, please still code review the PR and give feedback regardless if this will get merged this week or next.

Given the high number of Storage PRs we should start reviewing our own team's PRs to the storage/sdk as a first pass before tagging the storage team to lighten their PR load a bit more.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java [hunk 2]
```diff
@@ -300,7 +304,7 @@ public URL getContainerUrl() {
         if (!validateNoEtag(accessConditions.modifiedAccessConditions())) {
             // Throwing is preferred to Single.error because this will error out immediately instead of waiting until
             // subscription.
-            throw new UnsupportedOperationException("ETag access conditions are not supported for this API.");
+            logger.logAndThrow(new UnsupportedOperationException(ETAG_NOT_ALLOWED_MESSAGE));
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java]: It is not really needed to add a new message field. No need to be consistent. If you think the entire block is duplicate, it is much better to have a helper method.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java [hunk 3]
```diff
@@ -362,8 +366,7 @@ public URL getContainerUrl() {
             || accessConditions.modifiedAccessConditions().ifUnmodifiedSince() != null) {
             // Throwing is preferred to Single.error because this will error out immediately instead of waiting until
             // subscription.
-            throw new UnsupportedOperationException(
-                "If-Modified-Since is the only HTTP access condition supported for this API");
+            logger.logAndThrow(new UnsupportedOperationException("If-Modified-Since is the only HTTP access condition supported for this API"));
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java]: "return null"

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java [hunk 4]
```diff
@@ -658,7 +661,7 @@ public URL getContainerUrl() {
     private Mono<ContainersListBlobHierarchySegmentResponse> listBlobsHierarchySegment(String marker, String delimiter, ListBlobsOptions options) {
         options = options == null ? new ListBlobsOptions() : options;
         if (options.details().snapshots()) {
-            throw new UnsupportedOperationException("Including snapshots in a hierarchical listing is not supported.");
+            logger.logAndThrow(new UnsupportedOperationException("Including snapshots in a hierarchical listing is not supported."));
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ContainerAsyncClient.java]: return null, Also applied to some other places.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ServiceSASSignatureValues.java [hunk 5]
```diff
@@ -232,20 +234,18 @@ public ServiceSASSignatureValues canonicalName(String canonicalName) {
...
+            URL url = new URL(urlString);
+            StringBuilder canonicalName = new StringBuilder("/blob");
+            canonicalName.append('/').append(accountName).append(url.getPath());
+            this.canonicalName = canonicalName.toString();
         } catch (MalformedURLException e) {
-            throw new RuntimeException(e);
+            logger.logAndThrow(new IllegalArgumentException(e));
         }
 
-        StringBuilder canonicalName = new StringBuilder("/blob");
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ServiceSASSignatureValues.java]: The statements here does not have the chance of throwing MalformatedURLException. It is better to leave outside.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ServiceSASSignatureValues.java [hunk 6]
```diff
@@ -377,7 +377,7 @@ public SASQueryParameters generateSASQueryParameters(SharedKeyCredential sharedK
         try {
             signature = sharedKeyCredentials.computeHmac256(stringToSign);
         } catch (InvalidKeyException e) {
-            throw new Error(e); // The key should have been validated by now. If it is no longer valid here, we fail.
+            logger.logAndThrow(new IllegalStateException(e)); // The key should have been validated by now. If it is no longer valid here, we fail.
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ServiceSASSignatureValues.java]: IllegalStateException seems not very relevant.

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/FileClientBuilder.java [hunk 7]
```diff
@@ -121,7 +123,7 @@ public FileAsyncClient buildAsyncClient() {
         }
 
         if (sasTokenCredential == null && sharedKeyCredential == null) {
-            throw new IllegalArgumentException("Credentials are required for authorization");
+            logger.logAndThrow(new IllegalStateException("Credentials are required for authorization"));
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/FileClientBuilder.java]: IllegalStateException usually thrown when something wrong invoked in runtime. It is more like a RuntimeException

File: sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/ShareAsyncClient.java [hunk 8]
```diff
@@ -88,14 +91,15 @@
...
-     * @throws RuntimeException If the share is using a malformed URL.
+     * @throws IllegalStateException If the share is using a malformed URL.
      */
     public URL getShareUrl() {
         try {
             return new URL(azureFileStorageClient.getUrl());
         } catch (MalformedURLException e) {
-            throw new RuntimeException(String.format("Invalid URL on %s: %s" + getClass().getSimpleName(),
-                azureFileStorageClient.getUrl()), e);
+            logger.logAndThrow(new IllegalStateException(String.format("Invalid URL on %s: %s" + getClass().getSimpleName(),
```
Comment (sima-zhu) [inline sdk/storage/azure-storage-file/src/main/java/com/azure/storage/file/ShareAsyncClient.java]: Wrong usage of string format.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 9]
```diff
@@ -475,15 +479,16 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
...
+            return null;
         }
     }
 
     private void downloadToFileCleanup(AsynchronousFileChannel channel) {
         try {
             channel.close();
         } catch (IOException e) {
-            throw new UncheckedIOException(e);
+            logger.logAndThrow(new UncheckedIOException(e));
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java [hunk 10]
```diff
@@ -824,7 +829,7 @@ private void downloadToFileCleanup(AsynchronousFileChannel channel) {
         if (!(duration == -1 || (duration >= 15 && duration <= 60))) {
             // Throwing is preferred to Mono.error because this will error out immediately instead of waiting until
             // subscription.
-            throw new IllegalArgumentException("Duration must be -1 or between 15 and 60.");
+            logger.logAndThrow(new IllegalArgumentException("Duration must be -1 or between 15 and 60."));
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java [hunk 11]
```diff
@@ -341,7 +343,7 @@ public VoidResponse download(OutputStream stream, BlobRange range, ReliableDownl
                     try {
                         stream.write(bf.array());
                     } catch (IOException e) {
-                        throw new UncheckedIOException(e);
+                        logger.logAndThrow(new UncheckedIOException(e));
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 12]
```diff
@@ -253,7 +255,7 @@ public BlobClientBuilder endpoint(String endpoint) {
                 this.sharedKeyCredential = null;
             }
         } catch (MalformedURLException ex) {
-            throw new IllegalArgumentException("The Azure Storage Blob endpoint url is malformed.");
+            logger.logAndThrow(new IllegalArgumentException("The Azure Storage Blob endpoint url is malformed."));
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java [hunk 13]
```diff
@@ -363,7 +365,7 @@ public BlobClientBuilder connectionString(String connectionString) {
         String endpointSuffix = connectionKVPs.get(ENDPOINT_SUFFIX);
 
         if (ImplUtils.isNullOrEmpty(accountName) || ImplUtils.isNullOrEmpty(accountKey)) {
-            throw new IllegalArgumentException("Connection string must contain 'AccountName' and 'AccountKey'.");
+            logger.logAndThrow(new IllegalArgumentException("Connection string must contain 'AccountName' and 'AccountKey'."));
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobInputStream.java [hunk 14]
```diff
@@ -345,7 +348,7 @@ public int read(final byte[] b) throws IOException {
     @Override
     public int read(final byte[] b, final int off, final int len) throws IOException {
         if (off < 0 || len < 0 || len > b.length - off) {
-            throw new IndexOutOfBoundsException();
+            logger.logAndThrow(new IndexOutOfBoundsException());
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobInputStream.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobInputStream.java [hunk 15]
```diff
@@ -123,7 +126,7 @@
         this.accessCondition = accessCondition;
 
         if (blobRangeOffset < 0 || (blobRangeLength != null && blobRangeLength <= 0)) {
-            throw new IndexOutOfBoundsException();
+            logger.logAndThrow(new IndexOutOfBoundsException());
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobInputStream.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobInputStream.java [hunk 16]
```diff
@@ -447,7 +450,7 @@ public synchronized long skip(final long n) throws IOException {
         }
 
         if (n < 0 || this.currentAbsoluteReadPosition + n > this.streamLength + this.blobRangeOffset) {
-            throw new IndexOutOfBoundsException();
+            logger.logAndThrow(new IndexOutOfBoundsException());
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobInputStream.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobServiceAsyncClient.java [hunk 17]
```diff
@@ -235,7 +238,7 @@ public URL getAccountUrl() {
     public Mono<Response<UserDelegationKey>> getUserDelegationKey(OffsetDateTime start, OffsetDateTime expiry) {
         Utility.assertNotNull("expiry", expiry);
         if (start != null && !start.isBefore(expiry)) {
-            throw new IllegalArgumentException("`start` must be null or a datetime before `expiry`.");
+            logger.logAndThrow(new IllegalArgumentException("`start` must be null or a datetime before `expiry`."));
```
Comment (mssfang) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobServiceAsyncClient.java]: always end with 'return' statement after logger.logAndThrow()

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java (hunk 12 above)
Comment (joshfree) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClientBuilder.java]: @mssfang can you work with @alzimmermsft and make sure this starts failing CI very early in Preview 3?  We need to quickly get out o the world that code can pass CI but break the build after merge.

**Comment by alzimmermsft** [comment]
Changes to use `ClientLogger`'s exception logging is being handled in #4566

**Pull request closed without merging**